### PR TITLE
chore(skills): add profiling-capture skill

### PR DESCRIPTION
**Pull request by Yanko-7**
## Motivation

Add a **capture-only** XProf profiling skill for sgl-jax (JAX/TPU) under `.claude/skills/profiling-capture/`. It documents how to reliably *produce* trace artifacts (`*.xplane.pb` + `trace.json.gz`) for later analysis, avoiding common silent failures (cold runs, prefix-cache hits faking prefill, truncated/empty windows, LLO dump flags set after `import jax`). Analysis is out of scope.

## Modifications

Docs only — no runtime code changed. Adds three files:

- `SKILL.md` — path A/B/C selection (live server / offline batch / kernel microbench),
  capture-quality musts, handoff contract.
- `references/capture-recipes.md` — step definition, load shaping.
- `references/pallas-kernel-profiling.md` — Pallas/Mosaic LLO/HLO/XProf recipe + pitfalls.

## Accuracy Tests

N/A — documentation only, no model output changes.

## Benchmarking and Profiling

N/A — documentation only, no inference-speed impact.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

### CODE REVIEW

File: .claude/skills/profiling-capture/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+`output_dir` per request instead. Either way the dir must be writable on the host.
+
+## B — offline batch
+
+`bench_one_batch` takes the **same server args as `launch_server`**, so fill the model and
+parallelism from your own deployment context, then add the batch shape and `--profile`:
+
+```bash
+python -m sgl_jax.bench_one_batch \
+  --model-path <model> --tp-size <n> --device tpu ... \  # your launch_server config
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/SKILL.md]: This copy-paste command does not continue onto the next line. In POSIX shells, the backslash escapes the following space and `#` starts a comment, so line 85 is parsed as a separate command and fails with `--batch-size: command not found`. Please move the comment outside the command block, or make `\` the final character on the line.

File: .claude/skills/profiling-capture/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+| `--flush-cache` | clear the radix cache once before the run (see Capture quality) |
+| `--profile-by-stage` | write pre-split `prefill/` + `decode/` dirs (prefill unreliable) |
+| `--profile-stages prefill decode` | with `--profile-by-stage`, which stages to profile |
+
+**Need pre-split `prefill/` + `decode/` dirs** to compare the stages directly? Add
+`--profile-by-stage`, optionally with `--profile-stages prefill decode` — pass one stage to
+profile only that stage, omit for both.
+
+**Manual path:** `POST /start_profile` (body = `ProfileReqInput`, may include `output_dir`) →
+send requests → `POST /stop_profile` → `GET /profile_status` until `message == "idle"` (else
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/SKILL.md]: The manual flow polls the wrong response field. `/profile_status` returns `{"status":"idle"}` or `{"status":"in_progress"}`, not a `message` field, so a client following this instruction cannot observe completion. Please change this to `status == "idle"` (ideally with a copy-pasteable `curl | jq -r .status` example).

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    for _ in range(100):                      # a ns-scale kernel needs many iters or the window is ~empty
+        r = run_kernel()
+    jax.block_until_ready(r)                  # block_until_ready MUST be inside the block
+```
+
+Loop count: one call of a fast (single-digit-ns) kernel leaves too little in the window — a
+valid-but-empty trap. Loop until the region is milliseconds of device work (hundreds–thousands
+of iters for a tiny kernel); block **once at the end, still inside the block**.
+
+Name the phases with `jax.named_scope("w1_load" / "gemm" / "store_output")` so
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: This is not valid Python: it evaluates `"w1_load" / "gemm" / "store_output"` before calling `jax.named_scope`, which raises `TypeError`. Please show the actual context-manager form (for example, `with jax.named_scope("gemm"):` around the relevant operations) rather than slash-separated illustrative names inside a call.

File: .claude/skills/profiling-capture/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+$SGLANG_JAX_PROFILER_DIR/            # server-side; default /tmp
+    plugins/profile/<run>/<host>.xplane.pb           # default: single tree (+ .trace.json.gz)
+# with --profile-by-stage: split into prefill/ and decode/ subtrees instead
+    prefill/plugins/profile/<run>/<host>.xplane.pb
+    decode/plugins/profile/<run>/<host>.xplane.pb
+```
+
+## Handoff
+
+Do not analyze — hand the trace to the analysis skill. The dir containing `plugins/profile` is
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/SKILL.md]: The handoff is not currently actionable: this repository does not contain the separate analysis skill referenced here. Please name and link the expected skill if it exists elsewhere, or explicitly state that analysis is not bundled and provide the manual XProf/TensorBoard next step.

File: .claude/skills/profiling-capture/SKILL.md (hunk 1 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/SKILL.md]: Fixed.

File: .claude/skills/profiling-capture/SKILL.md (hunk 2 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/SKILL.md]: changed to status and added curl example.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md (hunk 3 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: Fixed — now uses the real context-manager form: one with jax.named_scope("..."): block per phase inside run_kernel(). Also added @jax.jit.

File: .claude/skills/profiling-capture/SKILL.md (hunk 4 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/SKILL.md]: Fixed — Handoff now states no analysis skill is bundled, and adds the manual step: TensorBoard profile plugin (or XProf) on the dir containing plugins/profile.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    with jax.named_scope("gemm"):
+        acc = matmul(x, w1)
+    with jax.named_scope("store_output"):
+        store(acc)
+    return acc
+
+opts = jax.profiler.ProfileOptions()          # add counter config only for counters mode
+for _ in range(3):                            # warm up so compile/autotune is outside the trace
+    jax.block_until_ready(run_kernel())
+with jax.profiler.trace(f"{OUT}/xprof"):      # trace lands beside the LLO/HLO dumps under $OUT
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: Looks like `OUT` got dropped when this example was rewritten. `OUT=/tmp/pallas-profile` in the shell block above is only a shell variable, while this snippet needs a Python variable here, so this hits `NameError` even after the kernel placeholders are filled in. Could we define `OUT` in the Python block or read it from `os.environ`?

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    # wrap each phase in its own named_scope context manager
+    with jax.named_scope("w1_load"):
+        w1 = load_weights()
+    with jax.named_scope("gemm"):
+        acc = matmul(x, w1)
+    with jax.named_scope("store_output"):
+        store(acc)
+    return acc
+
+opts = jax.profiler.ProfileOptions()          # add counter config only for counters mode
```
Comment (cjx0709) [inline .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: Small follow-up: `opts` is currently unused. If someone adds the suggested counter configuration there, it never reaches the profiler. Could we either remove this line for the plain-trace example, or pass it through as `jax.profiler.trace(..., profiler_options=opts)`? `bench_v2.py` already uses that pattern.

Comment (Yanko-7) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: replaced with jax.profiler.trace(..., profiler_options=opts).

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md (hunk 5 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: we define OUT in Python code.

File: .claude/skills/profiling-capture/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+| running HTTP server | **A** — live capture (non-stage default) |
+| one batch shape, no server | **B** — `bench_one_batch --profile` |
+| one kernel | **C** — kernel microbench |
+
+## A — live server
+
+`bench_serving --profile` is the recommended default: it warms up, arms the profiler, drives
+a real dataset workload, then stops — **one trace over the whole run**. A single window
+reliably captures device ops for **both** stages; separating prefill from decode is an
+analysis-time step (out of scope here).
```
Comment (JamesBrianD) [inline .claude/skills/profiling-capture/SKILL.md:41]: For the live server, need to configure sglang-jax server so that it does not disable precompilation.

File: .claude/skills/profiling-capture/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+  `.xplane.pb` file itself):
+
+  ```bash
+  pip install tensorboard tensorboard-plugin-profile
+  tensorboard --logdir /tmp/myprofile   # open http://localhost:6006 → PROFILE tab
+  ```
+
+- Or load the `*.xplane.pb` directly in the standalone [XProf](https://github.com/openxla/xprof) viewer.
+
+Either way, the dir containing `plugins/profile` is what the tool consumes. Always report:
```
Comment (JamesBrianD) [inline .claude/skills/profiling-capture/SKILL.md]: https://openxla.org/xprof/jax_profiling
It is now officially recommended to use xprof by passing the full logdir to it; please prioritize this approach.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+--xla_enable_custom_call_region_trace=true \
+--xla_xprof_register_llo_debug_info=true \
+--xla_mosaic_dump_to=$OUT/compiler/llo"
+export XLA_FLAGS="${XLA_FLAGS:+$XLA_FLAGS }\
+--xla_dump_to=$OUT/compiler/hlo --xla_dump_hlo_as_text --xla_dump_hlo_as_proto"
+```
+
+- `--xla_mosaic_dump_to` → Pallas/Mosaic LLO
+- `--xla_enable_custom_call_region_trace=true` → custom-call regions enter the trace
+- `--xla_xprof_register_llo_debug_info=true` → XProf can align custom-call ↔ LLO
```
Comment (JamesBrianD) [inline .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: https://openxla.org/xprof/custom_call_profiling
Note that these two options should be enabled only as needed, as they increase profile size and collection overhead.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+## Three artifact types
+
+| Artifact | When | Path |
+| --- | --- | --- |
+| Mosaic **LLO** dump | compile time | `compiler/llo/` |
+| HLO / MLIR dump | compile time | `compiler/hlo/` |
+| **XProf** trace / xplane | run time | `plugins/profile/<ts>/` |
+
+Having an LLO dump does **not** mean you captured a runtime trace — a failed
+compile can still leave LLO/HLO but no valid `.xplane.pb`.
```
Comment (JamesBrianD) [inline .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: The four capture modes provide different kinds of evidence and have different costs. They should not all be enabled by default. They can be combined, but each additional mode should be explicitly selected based on the profiling question.</p><div><div>
Mode | What it provides | How to enable it | Best suited for
-- | -- | -- | --
Plain trace | JAX/TPU runtime timeline, kernel duration, and named scopes | jax.profiler.trace(...) | “Which kernel is slow, and how long does it take?”
LLO utilization | LLO bundle utilization inside a Pallas custom call | Add the two custom-call profiling flags to LIBTPU_INIT_ARGS | “How are MXU, DMA, and wait time distributed inside the kernel?”
Compiler dump | Compile-time HLO, MLIR, and Mosaic LLO files | --xla_mosaic_dump_to and --xla_dump_to | “What instructions and layouts did the compiler generate?”
Periodic counters | Fine-grained, time-series hardware counters on Ironwood/TPU v7x | ProfileOptions.advanced_configuration | “Why is the MXU idle? Are spills or other hardware bottlenecks occurring?”

</div></div><p>I suggest structuring the recipe as four independent, opt-in layers:</p><ol start="1"><li>Use a <strong>plain trace</strong> as the low-overhead default.</li><li>Add <strong>LLO utilization</strong> when custom-call bundle visibility is required.</li><li>Enable <strong>compiler dumps</strong> only when investigating lowering, layouts, or generated instructions.</li><li>Enable <strong>periodic counters</strong> only for focused Ironwood/v7x bottleneck analysis.</li></ol><p>These modes are composable, but enabling all of them unconditionally can substantially increase compilation time, profile size, and runtime overhead.</p></body></html>

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md (hunk 9 above)
Comment (JamesBrianD) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: “--xla_enable_custom_call_region_trace=true” This flag incurs a performance penalty; this needs to be noted in the `skills` section as well.

File: .claude/skills/profiling-capture/SKILL.md (hunk 7 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/SKILL.md:41]: It has been clarified that the live server retain the default startup precompilation.

File: .claude/skills/profiling-capture/SKILL.md (hunk 8 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/SKILL.md]: The XProf has been set as the preferred option.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md (hunk 9 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: Changed the two custom-call profiling flags to be enabled on demand, with notes on their profile size and collection overhead.

File: .claude/skills/profiling-capture/references/pallas-kernel-profiling.md (hunk 10 above)
Comment (Yanko-7) [reply .claude/skills/profiling-capture/references/pallas-kernel-profiling.md]: Refactor Pallas profiling into a four-layer structure with plain trace as the default and the other three modes independently opt-in.

**Review by JamesBrianD — APPROVED** [review]
